File: adapters/python/runeward/langchain_tools.py
"""LangChain tool wrappers around :class:`RunewardClient`.

LangChain is imported *lazily* inside :func:`make_runeward_tools` so that the
base ``runeward`` client keeps working without ``langchain`` installed. Install
the extra with ``pip install runeward[langchain]``.

Each returned tool converts governance outcomes into a short, model-readable
string rather than letting the exception escape, so an agent can reason about a
denial or approval gate instead of crashing. The messages deliberately spell out
the required behavior ("do not retry", "pause for a human").
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List

from .client import RunewardApprovalRequired, RunewardClient, RunewardDenied

if TYPE_CHECKING:  # only for type checkers; not evaluated at runtime
    from langchain_core.tools import BaseTool


def make_runeward_tools(client: RunewardClient) -> "List[BaseTool]":
    """Build a list of LangChain tools bound to ``client``.

    Returns ``StructuredTool`` instances covering the runeward tool surface:
    create/kill Citadel, shell, python, node, file read/write/list/search, and
    list-Conclave.
    """
    # Lazy import: keeps langchain optional for users of the bare client.
    try:
        from langchain_core.tools import StructuredTool
    except ImportError as exc:  # pragma: no cover - depends on optional extra
        raise ImportError(
            "LangChain is required for make_runeward_tools(). "
            "Install it with: pip install runeward[langchain]"
        ) from exc

    def _guard(fn):
        """Wrap a call so governance verdicts become model-friendly strings."""

        def wrapped(*args: Any, **kwargs: Any) -> str:
            try:
                result = fn(*args, **kwargs)
                return result if isinstance(result, str) else str(result)
            except RunewardDenied as denied:
                return (
                    f"DENIED by policy: {denied.reason}. "
                    "Do not retry this action; choose a different, allowed approach "
                    "or tell the human it was blocked."
                )
            except RunewardApprovalRequired as approval:
                return (
                    f"APPROVAL REQUIRED (approval_id={approval.approval_id}): "
                    f"{approval.reason or 'a human must sign off before this runs'}. "
                    "Pause the task and ask the human to approve or deny."
                )

        return wrapped

    @_guard
    def create_citadel(profile: str) -> str:
        return str(client.create_sandbox(profile))

    @_guard
    def shell(sandbox: str, command: List[str], workdir: str = "") -> str:
        return str(client.shell(sandbox, command, workdir))

    @_guard
    def python(sandbox: str, code: str) -> str:
        return str(client.python(sandbox, code))

    @_guard
    def node(sandbox: str, code: str) -> str:
        return str(client.node(sandbox, code))

    @_guard
    def read_file(sandbox: str, path: str) -> str:
        return client.read_file(sandbox, path)

    @_guard
    def write_file(sandbox: str, path: str, content: str) -> str:
        return f"wrote {client.write_file(sandbox, path, content)} bytes to {path}"

    @_guard
    def list_files(sandbox: str, path: str) -> str:
        return client.list_files(sandbox, path)

    @_guard
    def search_files(sandbox: str, query: str, path: str) -> str:
        return client.search_files(sandbox, query, path)

    @_guard
    def list_conclave() -> str:
        return str(client.list_approvals())

    @_guard
    def kill_citadel(sandbox: str) -> str:
        client.kill_sandbox(sandbox)
        return f"Citadel {sandbox} terminated"

    return [
        StructuredTool.from_function(
            func=create_citadel,
            name="runeward_create_citadel",
            description="Provision a governed Citadel from a runeward Charter (e.g. 'dev'). Returns the Citadel metadata including its id.",
        ),
        StructuredTool.from_function(
            func=shell,
            name="runeward_shell",
            description="Run a shell command (as an argv list, e.g. ['ls','-la']) in a Citadel. Returns verdict, exit_code, stdout, stderr.",
        ),
        StructuredTool.from_function(
            func=python,
            name="runeward_python",
            description="Run a Python code snippet inside the Citadel.",
        ),
        StructuredTool.from_function(
            func=node,
            name="runeward_node",
            description="Run a Node.js code snippet inside the Citadel.",
        ),
        StructuredTool.from_function(
            func=read_file,
            name="runeward_read_file",
            description="Read a file's contents from the Citadel.",
        ),
        StructuredTool.from_function(
            func=write_file,
            name="runeward_write_file",
            description="Write content to a file in the Citadel.",
        ),
        StructuredTool.from_function(
            func=list_files,
            name="runeward_list_files",
            description="List a directory in the Citadel.",
        ),
        StructuredTool.from_function(
            func=search_files,
            name="runeward_search_files",
            description="Search for a query string under a path in the Citadel.",
        ),
        StructuredTool.from_function(
            func=list_conclave,
            name="runeward_list_conclave",
            description="List pending human-in-the-loop Conclave requests.",
        ),
        StructuredTool.from_function(
            func=kill_citadel,
            name="runeward_kill_citadel",
            description="Tear down a Citadel when the task is finished.",
        ),
    ]

File: adapters/typescript/src/langchain-tools.ts
/**
 * LangChain.js tool wrappers around {@link RunewardClient}.
 *
 * `@langchain/core` and `zod` are **optional peer dependencies** — this module
 * imports them at call time inside {@link makeRunewardTools} so the core client
 * stays dependency-free. Install them where you use the tools:
 *
 * ```bash
 * npm install @langchain/core zod
 * ```
 *
 * Each tool converts governance verdicts into a short, model-readable string
 * (rather than throwing) so the model can reason about a denial or an approval
 * gate: a `deny` must not be retried blindly, and a `require-approval` is a hard
 * pause for a human.
 */

import { RunewardApprovalRequired, RunewardClient, RunewardDenied } from "./client.js";

/** Turn a governance error into a model-friendly instruction; rethrow otherwise. */
function describeGovernance(err: unknown): string {
  if (err instanceof RunewardDenied) {
    return (
      `DENIED by policy: ${err.reason}. Do not retry this action; choose a ` +
      `different, allowed approach or tell the human it was blocked.`
    );
  }
  if (err instanceof RunewardApprovalRequired) {
    return (
      `APPROVAL REQUIRED (approval_id=${err.approvalId}): ` +
      `${err.reason || "a human must sign off before this runs"}. ` +
      `Pause the task and ask the human to approve or deny.`
    );
  }
  throw err;
}

/** Run `fn`, converting governance verdicts to strings for the model. */
async function guarded<T>(fn: () => Promise<T>): Promise<string> {
  try {
    const result = await fn();
    return typeof result === "string" ? result : JSON.stringify(result);
  } catch (err) {
    return describeGovernance(err);
  }
}

/**
 * Build an array of LangChain.js `DynamicStructuredTool`s bound to `client`,
 * named to match the runeward MCP tools. Pass them to an agent (e.g.
 * `createReactAgent({ llm, tools })`) or bind them to a chat model with
 * `model.bindTools(tools)`.
 *
 * @example
 * ```ts
 * import { ChatOpenAI } from "@langchain/openai";
 * import { createReactAgent } from "@langchain/langgraph/prebuilt";
 * import { RunewardClient } from "@runeward/sdk";
 * import { makeRunewardTools } from "@runeward/sdk/langchain-tools";
 *
 * const tools = await makeRunewardTools(new RunewardClient());
 * const agent = createReactAgent({ llm: new ChatOpenAI({ model: "gpt-4o" }), tools });
 * ```
 */
export async function makeRunewardTools(client: RunewardClient) {
  // Dynamic imports keep `@langchain/core` and `zod` as optional peers.
  const { DynamicStructuredTool } = await import("@langchain/core/tools");
  const { z } = await import("zod");

  return [
    new DynamicStructuredTool({
      name: "runeward_create_sandbox",
      description:
        "Provision a governed sandbox from a runeward profile (e.g. 'dev'). Returns sandbox metadata including its id.",
      schema: z.object({
        profile: z.string().describe("Profile name, e.g. 'dev' or 'governed'."),
      }),
      func: async ({ profile }) => guarded(() => client.createSandbox(profile)),
    }),

    new DynamicStructuredTool({
      name: "runeward_shell",
      description:
        "Run a shell command (as an argv array, e.g. ['ls','-la']) in a sandbox. Returns verdict, exit_code, stdout, stderr.",
      schema: z.object({
        sandbox: z.string().describe("Sandbox id from create_sandbox."),
        command: z.array(z.string()).describe("argv array, e.g. ['ls','-la']."),
        workdir: z.string().optional().describe("Optional working directory."),
      }),
      func: async ({ sandbox, command, workdir }) =>
        guarded(() => client.shell(sandbox, command, workdir ?? "")),
    }),

    new DynamicStructuredTool({
      name: "runeward_python",
      description: "Run a Python code snippet inside the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        code: z.string().describe("Python source to execute."),
      }),
      func: async ({ sandbox, code }) => guarded(() => client.python(sandbox, code)),
    }),

    new DynamicStructuredTool({
      name: "runeward_node",
      description: "Run a Node.js code snippet inside the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        code: z.string().describe("JavaScript source to execute."),
      }),
      func: async ({ sandbox, code }) => guarded(() => client.node(sandbox, code)),
    }),

    new DynamicStructuredTool({
      name: "runeward_read_file",
      description: "Read a file's contents from the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        path: z.string().describe("File path to read."),
      }),
      func: async ({ sandbox, path }) => guarded(() => client.readFile(sandbox, path)),
    }),

    new DynamicStructuredTool({
      name: "runeward_write_file",
      description: "Write content to a file in the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        path: z.string().describe("File path to write."),
        content: z.string().describe("Content to write."),
      }),
      func: async ({ sandbox, path, content }) =>
        guarded(async () => `wrote ${await client.writeFile(sandbox, path, content)} bytes to ${path}`),
    }),

    new DynamicStructuredTool({
      name: "runeward_list_files",
      description: "List a directory in the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        path: z.string().describe("Directory path to list."),
      }),
      func: async ({ sandbox, path }) => guarded(() => client.listFiles(sandbox, path)),
    }),

    new DynamicStructuredTool({
      name: "runeward_search_files",
      description: "Search for a query string under a path in the sandbox.",
      schema: z.object({
        sandbox: z.string(),
        query: z.string().describe("Search query."),
        path: z.string().describe("Path to search under."),
      }),
      func: async ({ sandbox, query, path }) =>
        guarded(() => client.searchFiles(sandbox, query, path)),
    }),

    new DynamicStructuredTool({
      name: "runeward_list_approvals",
      description: "List pending human-in-the-loop approval requests.",
      schema: z.object({}),
      func: async () => guarded(() => client.listApprovals()),
    }),

    new DynamicStructuredTool({
      name: "runeward_kill_sandbox",
      description: "Tear down a sandbox when the task is finished.",
      schema: z.object({
        sandbox: z.string(),
      }),
      func: async ({ sandbox }) =>
        guarded(async () => {
          await client.killSandbox(sandbox);
          return `sandbox ${sandbox} terminated`;
        }),
    }),
  ];
}
